File: pyro_integration/workers/transcoder.py
"""
R-YARA Transcoder Worker

Worker implementation for YARA rule transcoding (codename conversion).
"""

import asyncio
import json
import subprocess
import tempfile
import os
import re
from typing import Dict, Any, List, Optional
from pathlib import Path

from .base import RYaraWorker
from ..shared.protocol import WorkerTask, TaskResult, TaskType


class TranscoderWorker(RYaraWorker):
    """
    Transcoder worker for R-YARA codename operations.

    Handles:
    - Encode rules with codenames (obfuscation)
    - Decode rules from codenames (deobfuscation)
    - Dictionary lookups
    - Batch transcoding operations
    """

    @property
    def worker_type(self) -> str:
        return "r-yara-transcoder"

    @property
    def capabilities(self) -> list:
        return [
            TaskType.TRANSCODE.value,
            TaskType.DICTIONARY_LOOKUP.value
        ]

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._cli_binary = self._find_ryara_cli()
        self._dictionary_cache: Dict[str, str] = {}
        self._reverse_cache: Dict[str, str] = {}

    def _find_ryara_cli(self) -> Optional[str]:
        """Find R-YARA CLI binary"""
        locations = [
            "r-yara-cli",
            str(Path(__file__).parent.parent.parent / "rust" / "target" / "release" / "r-yara-cli"),
            str(Path(__file__).parent.parent.parent / "rust" / "target" / "debug" / "r-yara-cli"),
            "/usr/local/bin/r-yara-cli",
        ]
        for loc in locations:
            try:
                result = subprocess.run(
                    [loc, "--help"],
                    capture_output=True,
                    timeout=5
                )
                if result.returncode == 0:
                    return loc
            except (subprocess.SubprocessError, FileNotFoundError):
                continue
        return None

    async def process_task(self, task: WorkerTask) -> TaskResult:
        """Process a transcoder task"""
        if task.task_type == TaskType.TRANSCODE:
            return await self._transcode(task)
        elif task.task_type == TaskType.DICTIONARY_LOOKUP:
            return await self._dictionary_lookup(task)
        else:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error=f"Unknown task type: {task.task_type}"
            )

    async def _transcode(self, task: WorkerTask) -> TaskResult:
        """Transcode YARA rules (encode/decode codenames)"""
        rule_content = task.payload.get("rule")
        direction = task.payload.get("direction", "encode")  # encode or decode

        if not rule_content:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error="Missing rule in payload"
            )

        try:
            if direction == "encode":
                result = await self._encode_rule(rule_content)
            else:
                result = await self._decode_rule(rule_content)

            return TaskResult(
                task_id=task.task_id,
                success=True,
                data={
                    "original": rule_content,
                    "transcoded": result["transcoded"],
                    "mappings": result["mappings"],
                    "direction": direction
                }
            )
        except Exception as e:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error=str(e)
            )

    async def _encode_rule(self, rule_content: str) -> Dict[str, Any]:
        """Encode a rule with codenames"""
        # If CLI is available, use it
        if self._cli_binary:
            return await self._encode_with_cli(rule_content)

        # Fallback to Python implementation
        return self._encode_with_python(rule_content)

    async def _decode_rule(self, rule_content: str) -> Dict[str, Any]:
        """Decode a rule from codenames"""
        # If CLI is available, use it
        if self._cli_binary:
            return await self._decode_with_cli(rule_content)

        # Fallback to Python implementation
        return self._decode_with_python(rule_content)

    async def _encode_with_cli(self, rule_content: str) -> Dict[str, Any]:
        """Encode using R-YARA CLI"""
        temp_file = tempfile.NamedTemporaryFile(
            mode='w',
            suffix='.yar',
            delete=False
        )
        temp_file.write(rule_content)
        temp_file.close()

        try:
            process = await asyncio.create_subprocess_exec(
                self._cli_binary, "encode", temp_file.name,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )
            stdout, stderr = await process.communicate()

            if process.returncode != 0:
                raise Exception(stderr.decode())

            result = json.loads(stdout.decode())
            return {
                "transcoded": result.get("encoded", ""),
                "mappings": result.get("mappings", {})
            }
        finally:
            os.unlink(temp_file.name)

    async def _decode_with_cli(self, rule_content: str) -> Dict[str, Any]:
        """Decode using R-YARA CLI"""
        temp_file = tempfile.NamedTemporaryFile(
            mode='w',
            suffix='.yar',
            delete=False
        )
        temp_file.write(rule_content)
        temp_file.close()

        try:
            process = await asyncio.create_subprocess_exec(
                self._cli_binary, "decode", temp_file.name,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )
            stdout, stderr = await process.communicate()

            if process.returncode != 0:
                raise Exception(stderr.decode())

            result = json.loads(stdout.decode())
            return {
                "transcoded": result.get("decoded", ""),
                "mappings": result.get("mappings", {})
            }
        finally:
            os.unlink(temp_file.name)

    def _encode_with_python(self, rule_content: str) -> Dict[str, Any]:
        """Encode using Python (fallback)"""
        # Simple encoding: replace identifiers with codenames
        mappings = {}
        transcoded = rule_content

        # Find rule names
        rule_pattern = r'rule\s+(\w+)'
        for match in re.finditer(rule_pattern, rule_content):
            original = match.group(1)
            if original not in mappings:
                codename = self._generate_codename(original)
                mappings[original] = codename

        # Find string identifiers
        string_pattern = r'\$(\w+)\s*='
        for match in re.finditer(string_pattern, rule_content):
            original = f"${match.group(1)}"
            if original not in mappings:
                codename = self._generate_codename(original)
                mappings[original] = codename

        # Apply mappings
        for original, codename in mappings.items():
            transcoded = transcoded.replace(original, codename)

        return {
            "transcoded": transcoded,
            "mappings": mappings
        }

    def _decode_with_python(self, rule_content: str) -> Dict[str, Any]:
        """Decode using Python (fallback)"""
        # Simple decoding: look up codenames in reverse cache
        mappings = {}
        transcoded = rule_content

        # Find potential codenames (format: R_XXXX or similar)
        codename_pattern = r'R_[A-Z0-9]{4,8}'
        for match in re.finditer(codename_pattern, rule_content):
            codename = match.group(0)
            if codename in self._reverse_cache:
                original = self._reverse_cache[codename]
                mappings[codename] = original
                transcoded = transcoded.replace(codename, original)

        return {
            "transcoded": transcoded,
            "mappings": mappings
        }

    def _generate_codename(self, identifier: str) -> str:
        """Generate a codename for an identifier"""
        import hashlib
        # Generate hash-based codename
        hash_val = hashlib.md5(identifier.encode()).hexdigest()[:6].upper()
        codename = f"R_{hash_val}"

        # Cache the mapping
        self._dictionary_cache[identifier] = codename
        self._reverse_cache[codename] = identifier

        return codename

    async def _dictionary_lookup(self, task: WorkerTask) -> TaskResult:
        """Look up entries in the dictionary"""
        query = task.payload.get("query")
        lookup_type = task.payload.get("type", "codename")  # codename or symbol

        if not query:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error="Missing query in payload"
            )

        try:
            # Try CLI first
            if self._cli_binary:
                result = await self._lookup_with_cli(query, lookup_type)
            else:
                result = self._lookup_with_cache(query, lookup_type)

            return TaskResult(
                task_id=task.task_id,
                success=True,
                data=result
            )
        except Exception as e:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error=str(e)
            )

    async def _lookup_with_cli(self, query: str, lookup_type: str) -> Dict[str, Any]:
        """Look up using R-YARA CLI"""
        process = await asyncio.create_subprocess_exec(
            self._cli_binary, "lookup", query,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE
        )
        stdout, stderr = await process.communicate()

        if process.returncode != 0:
            return {"found": False, "query": query}

        try:
            return json.loads(stdout.decode())
        except json.JSONDecodeError:
            return {
                "found": True,
                "query": query,
                "result": stdout.decode().strip()
            }

    def _lookup_with_cache(self, query: str, lookup_type: str) -> Dict[str, Any]:
        """Look up in local cache"""
        if lookup_type == "codename":
            if query in self._reverse_cache:
                return {
                    "found": True,
                    "codename": query,
                    "symbol": self._reverse_cache[query]
                }
        else:
            if query in self._dictionary_cache:
                return {
                    "found": True,
                    "symbol": query,
                    "codename": self._dictionary_cache[query]
                }

        return {"found": False, "query": query}

    async def load_dictionary(self, dictionary_path: str):
        """Load dictionary from file"""
        try:
            with open(dictionary_path, 'r') as f:
                data = json.load(f)

            for entry in data.get("entries", []):
                symbol = entry.get("symbol")
                codename = entry.get("codename")
                if symbol and codename:
                    self._dictionary_cache[symbol] = codename
                    self._reverse_cache[codename] = symbol
        except Exception as e:
            raise Exception(f"Failed to load dictionary: {e}")


def create_transcoder_worker(**kwargs) -> TranscoderWorker:
    """Factory function to create a transcoder worker"""
    return TranscoderWorker(**kwargs)

File: rust/r-yara-pyro/src/workers/transcoder.rs
//! Transcoder worker implementation
//!
//! Handles YARA rule transcoding (codename encoding/decoding) and dictionary operations.

use std::collections::HashMap;
use std::sync::Arc;
use std::time::Instant;
use tokio::sync::RwLock;
use regex::Regex;

use crate::protocol::{TaskResult, TaskType, WorkerTask};

use super::base::{BaseWorker, WorkerState, WorkerStats};

/// Transcoder worker for codename operations
pub struct TranscoderWorker {
    base: BaseWorker,
    /// Symbol to codename mapping
    dictionary: Arc<RwLock<HashMap<String, String>>>,
    /// Codename to symbol reverse mapping
    reverse_dictionary: Arc<RwLock<HashMap<String, String>>>,
}

impl TranscoderWorker {
    /// Create a new transcoder worker
    pub fn new() -> Self {
        let base = BaseWorker::new(
            "r-yara-transcoder",
            vec![TaskType::Transcode, TaskType::DictionaryLookup],
        );

        Self {
            base,
            dictionary: Arc::new(RwLock::new(HashMap::new())),
            reverse_dictionary: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    /// Get worker ID
    pub fn worker_id(&self) -> &str {
        &self.base.worker_id
    }

    /// Get worker type
    pub fn worker_type(&self) -> &str {
        &self.base.worker_type
    }

    /// Get capabilities
    pub fn capabilities(&self) -> Vec<TaskType> {
        self.base.capabilities.clone()
    }

    /// Get current state
    pub async fn state(&self) -> WorkerState {
        self.base.get_state().await
    }

    /// Get statistics
    pub async fn stats(&self) -> WorkerStats {
        self.base.get_stats().await
    }

    /// Process a task
    pub async fn process_task(&self, task: WorkerTask) -> TaskResult {
        let start = Instant::now();
        self.base.task_started(&task).await;

        let result = match task.task_type {
            TaskType::Transcode => self.transcode(&task).await,
            TaskType::DictionaryLookup => self.dictionary_lookup(&task).await,
            _ => TaskResult::failure(&task.task_id, "Unsupported task type"),
        };

        let elapsed = start.elapsed().as_millis() as u64;
        self.base
            .task_completed(&task.task_id, result.success, elapsed)
            .await;

        result.with_execution_time(elapsed)
    }

    /// Transcode a YARA rule
    async fn transcode(&self, task: &WorkerTask) -> TaskResult {
        let rule = match task.payload.get("rule") {
            Some(serde_json::Value::String(r)) => r.clone(),
            _ => return TaskResult::failure(&task.task_id, "Missing rule in payload"),
        };

        let direction = task
            .payload
            .get("direction")
            .and_then(|d| d.as_str())
            .unwrap_or("encode");

        let (transcoded, mappings) = if direction == "encode" {
            self.encode_rule(&rule).await
        } else {
            self.decode_rule(&rule).await
        };

        TaskResult::success(
            &task.task_id,
            serde_json::json!({
                "original": rule,
                "transcoded": transcoded,
                "mappings": mappings,
                "direction": direction
            }),
        )
    }

    /// Encode a rule with codenames
    async fn encode_rule(&self, rule: &str) -> (String, HashMap<String, String>) {
        let mut transcoded = rule.to_string();
        let mut mappings = HashMap::new();

        // Find rule names
        let rule_pattern = Regex::new(r"rule\s+(\w+)").unwrap();
        for cap in rule_pattern.captures_iter(rule) {
            let original = cap.get(1).unwrap().as_str();
            if !mappings.contains_key(original) {
                let codename = self.generate_codename(original).await;
                mappings.insert(original.to_string(), codename.clone());
            }
        }

        // Find string identifiers
        let string_pattern = Regex::new(r"\$(\w+)\s*=").unwrap();
        for cap in string_pattern.captures_iter(rule) {
            let original = format!("${}", cap.get(1).unwrap().as_str());
            if !mappings.contains_key(&original) {
                let codename = self.generate_codename(&original).await;
                mappings.insert(original.clone(), codename.clone());
            }
        }

        // Apply mappings
        for (original, codename) in &mappings {
            transcoded = transcoded.replace(original, codename);
        }

        (transcoded, mappings)
    }

    /// Decode a rule from codenames
    async fn decode_rule(&self, rule: &str) -> (String, HashMap<String, String>) {
        let mut transcoded = rule.to_string();
        let mut mappings = HashMap::new();

        // Find potential codenames (R_XXXXXX format)
        let codename_pattern = Regex::new(r"R_[A-F0-9]{6}").unwrap();
        let reverse_dict = self.reverse_dictionary.read().await;

        for mat in codename_pattern.find_iter(rule) {
            let codename = mat.as_str();
            if let Some(original) = reverse_dict.get(codename) {
                mappings.insert(codename.to_string(), original.clone());
                transcoded = transcoded.replace(codename, original);
            }
        }

        (transcoded, mappings)
    }

    /// Generate a codename for an identifier
    async fn generate_codename(&self, identifier: &str) -> String {
        use std::collections::hash_map::DefaultHasher;
        use std::hash::{Hash, Hasher};

        let mut hasher = DefaultHasher::new();
        identifier.hash(&mut hasher);
        let hash = hasher.finish();
        let codename = format!("R_{:06X}", hash as u32 & 0xFFFFFF);

        // Cache the mapping
        let mut dict = self.dictionary.write().await;
        dict.insert(identifier.to_string(), codename.clone());
        drop(dict);

        let mut reverse = self.reverse_dictionary.write().await;
        reverse.insert(codename.clone(), identifier.to_string());

        codename
    }

    /// Dictionary lookup
    async fn dictionary_lookup(&self, task: &WorkerTask) -> TaskResult {
        let query = match task.payload.get("query") {
            Some(serde_json::Value::String(q)) => q,
            _ => return TaskResult::failure(&task.task_id, "Missing query in payload"),
        };

        let lookup_type = task
            .payload
            .get("type")
            .and_then(|t| t.as_str())
            .unwrap_or("codename");

        let result = if lookup_type == "codename" {
            // Look up codename -> symbol
            let reverse = self.reverse_dictionary.read().await;
            if let Some(symbol) = reverse.get(query) {
                serde_json::json!({
                    "found": true,
                    "codename": query,
                    "symbol": symbol
                })
            } else {
                serde_json::json!({
                    "found": false,
                    "query": query
                })
            }
        } else {
            // Look up symbol -> codename
            let dict = self.dictionary.read().await;
            if let Some(codename) = dict.get(query) {
                serde_json::json!({
                    "found": true,
                    "symbol": query,
                    "codename": codename
                })
            } else {
                serde_json::json!({
                    "found": false,
                    "query": query
                })
            }
        };

        TaskResult::success(&task.task_id, result)
    }

    /// Load dictionary from JSON
    pub async fn load_dictionary(&self, data: &str) -> Result<usize, String> {
        let entries: Vec<serde_json::Value> =
            serde_json::from_str(data).map_err(|e| e.to_string())?;

        let mut dict = self.dictionary.write().await;
        let mut reverse = self.reverse_dictionary.write().await;
        let mut count = 0;

        for entry in entries {
            if let (Some(symbol), Some(codename)) = (
                entry.get("symbol").and_then(|s| s.as_str()),
                entry.get("codename").and_then(|c| c.as_str()),
            ) {
                dict.insert(symbol.to_string(), codename.to_string());
                reverse.insert(codename.to_string(), symbol.to_string());
                count += 1;
            }
        }

        Ok(count)
    }

    /// Get dictionary statistics
    pub async fn dictionary_stats(&self) -> serde_json::Value {
        let dict = self.dictionary.read().await;
        serde_json::json!({
            "total_entries": dict.len(),
            "status": if dict.is_empty() { "empty" } else { "loaded" }
        })
    }
}

impl Default for TranscoderWorker {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_transcoder_worker_creation() {
        let worker = TranscoderWorker::new();
        assert_eq!(worker.worker_type(), "r-yara-transcoder");
    }

    #[tokio::test]
    async fn test_encode_rule() {
        let worker = TranscoderWorker::new();
        let rule = r#"rule test_rule { strings: $a = "test" condition: $a }"#;
        let (transcoded, mappings) = worker.encode_rule(rule).await;

        assert!(!mappings.is_empty());
        assert!(!transcoded.contains("test_rule"));
    }

    #[tokio::test]
    async fn test_codename_generation() {
        let worker = TranscoderWorker::new();
        let codename = worker.generate_codename("test").await;
        assert!(codename.starts_with("R_"));
        assert_eq!(codename.len(), 8); // R_ + 6 hex chars
    }

    #[tokio::test]
    async fn test_dictionary_roundtrip() {
        let worker = TranscoderWorker::new();

        // Generate codename
        let codename = worker.generate_codename("my_symbol").await;

        // Look up by codename
        let reverse = worker.reverse_dictionary.read().await;
        assert_eq!(reverse.get(&codename), Some(&"my_symbol".to_string()));
    }
}
